### PR TITLE
Two tricks and a pitfall

### PR DESCRIPTION
**Pull request by phackstock**
With a little bit of delay, here's my code and tell from last week's message meeting.

### CODE REVIEW

**Comment by danielhuppmann** [comment]
I suggest that @LauWien reviews this PR instead of me

File: 2021-08-04-two-tricks-one-pitfall.ipynb [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Two tricks one pitfall \n",
+    "## Trick #1: The for-else-loop\n",
+    "\n",
+    "Python allows a for-else-loop. This is used for for-loops with a breaking condition to check whether the breaking condition was met or the loop ran through.\n",
+    "\n",
+    "### Example:\n",
+    "We have a list of models `models = ['Message', 'GCAM', 'REMIND']` and want to find out if any model name is shorter than 4 characters. If we find one, we want to print the first one we find. If we do not fine one, we want to say that we came up empty handed. \n",
```
Comment (LauWien) [inline 2021-08-04-two-tricks-one-pitfall.ipynb]:
```suggestion
    "We have a list of models `models = ['MESSAGE', 'GCAM', 'REMIND']` and want to find out if any model name is shorter than 4 characters. If we find one, we want to print the first one we find. If we do not fine one, we want to say that we came up empty handed. \n",
```

File: 2021-08-04-two-tricks-one-pitfall.ipynb [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "No model shorter than 4 characters found.\n"
+     ]
+    }
+   ],
+   "source": [
+    "models = ['Message', 'GCAM', 'REMIND']\n",
```
Comment (LauWien) [inline 2021-08-04-two-tricks-one-pitfall.ipynb]:
```suggestion
    "models = ['MESSAGE', 'GCAM', 'REMIND']\n",
```

File: 2021-08-04-two-tricks-one-pitfall.ipynb [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+    "print(f\"{a:c}\") # ascii\n",
+    "print(f\"{a:o}\") # octal\n",
+    "print(f\"{a:010b}\") # combined with padding, padding with 0 to a total of 10 digits"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## The Pitfall: Mutable default Arguments\n",
```
Comment (LauWien) [inline 2021-08-04-two-tricks-one-pitfall.ipynb]:
```suggestion
    "## The Pitfall: Mutable default arguments\n",
```

File: 2021-08-04-two-tricks-one-pitfall.ipynb [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+    "\n",
+    "The python interpreter evaluates default argument only **once** at the beginning of the interpretation process.\n",
```
Comment (LauWien) [inline 2021-08-04-two-tricks-one-pitfall.ipynb]:
```suggestion
    "The python interpreter evaluates default arguments only **once** at the beginning of the interpretation process.\n",
```

File: 2021-08-04-two-tricks-one-pitfall.ipynb [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+   "source": [
+    "# really simple function, appends a variable x to a list l \n",
+    "# the list l is optional, if we don't provide one, the function\n",
+    "# takes an empty one.\n",
+    "def add_to_list(x, l=[]):\n",
+    "    l.append(x)\n",
+    "    return l\n",
+    "\n",
+    "print(add_to_list(1)) # we would expect [1], we get [1]\n",
+    "print(add_to_list(2)) # we would expect [2], we get [1, 2]"
```
Comment (LauWien) [inline 2021-08-04-two-tricks-one-pitfall.ipynb]:
```suggestion
    "print(add_to_list(5)) # we would expect [5], we get [1, 5]"
```

File: 2021-08-04-two-tricks-one-pitfall.ipynb [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+    "    l.append(x)\n",
+    "    return l\n",
+    "\n",
+    "def print_time(dt = None):\n",
+    "    if dt is None:\n",
+    "        dt = datetime.now()\n",
+    "    print(dt)\n",
+    "\n",
+    "print(add_to_list(1))\n",
+    "print(add_to_list(2))\n",
```
Comment (LauWien) [inline 2021-08-04-two-tricks-one-pitfall.ipynb:319]:
```suggestion
    "print(add_to_list(5))\n",
```

**Comment by phackstock** [comment]
That's exactly what I wanted to do but for some reason GitHub didn't allow to select her as a reviewer. It probably has something to do with the access rights, I'll check

File: 2021-08-04-two-tricks-one-pitfall.ipynb [hunk 7]
```diff
@@ -0,0 +1,361 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Two tricks one pitfall \n",
```
Comment (LauWien) [inline 2021-08-04-two-tricks-one-pitfall.ipynb]:
```suggestion
    "# Two tricks and a pitfall \n",
```

File: README.md [hunk 8]
```diff
@@ -22,3 +22,4 @@ Please also make sure that the `requirements.txt` is kept up to date.
 - 2021-05-26 (@phackstock): Type Hints
 - 2021-06-09 (@phackstock): fnmatch
 - 2021-06-16 (@danielhuppmann): Using pyam to compute derived indicators
+- 2021-08-04 (@phackstock): Two tricks one pitfall
```
Comment (LauWien) [inline README.md]:
```suggestion
- 2021-08-04 (@phackstock): Two tricks and a pitfall
```

**Review by LauWien — CHANGES_REQUESTED** [review]
Nice! Just some suggestions :)

File: 2021-08-04-two-tricks-one-pitfall.ipynb (hunk 5 above)
Comment (phackstock) [reply 2021-08-04-two-tricks-one-pitfall.ipynb]: Is there a special reason to use 5 instead of 2?

Comment (LauWien) [reply 2021-08-04-two-tricks-one-pitfall.ipynb]: :D yes, sorry for not explaining. One could think the behavior origins from a loop with stepsize 1, which then produces the list [1, 2], if you insert 2 and not just beeing the mutable list. But I guess it's just me overthinking.

Comment (phackstock) [reply 2021-08-04-two-tricks-one-pitfall.ipynb]: I'm afraid I don't follow, there is no loop in `add_to_list`. But I can change it to a `5` no problem, that might direct people away from thinking that it has anything to do with 1 and 2 being consecutive numbers. Now that I think about it we should probably do something like:

``` python
print(add_to_list(1)) # we would expect [1], we get [1]\n",
print(add_to_list('hello')) # we would expect ['hello'], we get [1, 'hello']"
```

Comment (LauWien) [reply 2021-08-04-two-tricks-one-pitfall.ipynb]: Okay, great, that looks awesome. Sorry for the confusion. I guess we can then squash and merge.

**Pull request merged**